_docs/rfc/for_each_iteration.md

```diff
@@ -0,0 +1,218 @@
+
+- `ParseConfigFile`, `ParseConfigString`, `PartialParseConfigFile`, and `PartialParseConfigString` need to be updated to
+  return list of `TerragruntConfig` structs, as opposed to a single one.
+- Each caller of parsing functions need to accordingly handle the list of `TerragruntConfig` objects. As a part of this,
+  the caller should loop through each config to run the commands in the context of each iteration of the expanded loop.
+- The terragrunt workspace logic needs to be updated to ensure a separate module folder is created for each loop
+  iteration.
+- The parsing functions need to be updated to first parse the `for_each` attribute. Then, the parser should iterate each
+  item in the list and set the `each` variable accordingly as it parses the rest of the config. This means the parsing
+  order is now adjusted as:
+    - `for_each` and expand loop.
+    - `locals`
+    - `include`
+    - `dependency`
+    - Everything else
+- Support `merge_parent` in `remote_state`, which ultimately adjusts how `mergeConfigWithIncludedConfig` merges the
+  child `remote_state` config with the parent.
+
+
+##### Pros and Cons of approach
+
+**Pros**:
+
+- Parallels `for_each` in Terraform. In the examples I used `list`, but the implementation details can mimic that of
+  Terraform (requiring `set` or `map`). That way, the syntax closely resembles an existing construct and thus it will be
+  easier to pick up.
+- Limited changes are necessary in the user config to add in iteration.
+- The config parser implementation is relatively simple.
+
+**Cons**:
+
+- The implementation of the runners can become complex, especially in the dependency detection for the `xxx-all`
+  commands. If `dependency` blocks are parameterized by looping, this can be hairy to unpack and expand into the
+  dependency graph if the terragrunt files don't exist.
+- The logic for expanding out the loop for the terragrunt cache is complex as it is not immediately obvious what the
+  folder structure should be. Should it be a hash of the iteration value appended to the working path (e.g
+  `.terragrunt-cache/HASH_OF_EACH_VALUE/path/to/workspace`)? Or should it be one cache per iteration value (e.g
+  `HASH_OF_EACH_VALUE/.terragrunt-cache/path/to/workspace`)?
+- Implementation requires a major refactor in the calling logic post parsing.
+- It is not immediately obvious how to handle removing an item in the `for_each` list. A naive approach would not run
+  `terragrunt destroy` on that module. Additionally, as stated, it is impossible to run the command on a single item in
+  the list and thus you can't destroy just the items you want to remove without doing complex dancing (e.g update the
+  `for_each` list to only reference the items you want to destroy, then run `terragrunt destroy` to destroy just those
+  items, and then restore the list to the version with the items removed).
+
+### Option 3: scaffolding tool that code gens live config using a template
```

Comment:
@bwhaley : After contemplating about this feature, I am leaning towards this approach for our purposes. The implementation for supporting iteration in Terragrunt has many nuances that suggest it is not a great idea.

Better to wait for the official solution, working around it using code generation. Code generation appears to be the lesser evil of a workaround.